File: app/src/google/java/com/apps/adrcotfas/goodtime/upgrade/BillingHelper.java
package com.apps.adrcotfas.goodtime.upgrade;

import android.content.Context;
import android.content.Intent;

import com.anjlab.android.iab.v3.BillingProcessor;
import com.anjlab.android.iab.v3.TransactionDetails;
import com.apps.adrcotfas.goodtime.Main.IBillingHelper;
import com.apps.adrcotfas.goodtime.Settings.PreferenceHelper;
import com.apps.adrcotfas.goodtime.R;
import com.apps.adrcotfas.goodtime.Util.Constants;

import androidx.annotation.NonNull;

public class BillingHelper implements IBillingHelper, BillingProcessor.IBillingHandler  {

    private BillingProcessor mBillingProcessor;

    public BillingHelper(Context context) {
        mBillingProcessor = BillingProcessor.newBillingProcessor(
                context,
                context.getString(R.string.licence_key),
                context.getString(R.string.merchant_id),  this);
        mBillingProcessor.initialize();
    }

    @Override
    public void onProductPurchased(@NonNull String productId, TransactionDetails details) {
        // do nothing here
    }

    @Override
    public void onPurchaseHistoryRestored() {
        boolean found = false;
        for(String sku : mBillingProcessor.listOwnedProducts()) {
            if (sku.equals(Constants.sku)) {
                found = true;
                break;
            }
        }
        PreferenceHelper.setPro(found);
    }

    @Override
    public void onBillingError(int errorCode, Throwable error) {
        // do nothing here
    }

    @Override
    public void onBillingInitialized() {
        mBillingProcessor.loadOwnedPurchasesFromGoogle();
        if (mBillingProcessor.isPurchased(Constants.sku)) {
            PreferenceHelper.setPro(true);
        }
    }

    @Override
    public void refresh() {
        if (mBillingProcessor != null) {
            mBillingProcessor.loadOwnedPurchasesFromGoogle();
        }
    }

    @Override
    public void release() {
        if (mBillingProcessor != null)
            mBillingProcessor.release();
    }

    @Override
    public boolean handleActivityResult(int requestCode, int resultCode, Intent data) {
        if (mBillingProcessor != null) {
            mBillingProcessor.loadOwnedPurchasesFromGoogle();
            if (mBillingProcessor.isPurchased(Constants.sku)) {
                PreferenceHelper.setPro(true);
            }
            return mBillingProcessor.handleActivityResult(requestCode, resultCode, data);
        } else {
            return false;
        }
    }
}

File: app/src/google/java/com/apps/adrcotfas/goodtime/upgrade/BillingHelper.kt
package com.apps.adrcotfas.goodtime.upgrade

import android.content.Context
import com.apps.adrcotfas.goodtime.main.IBillingHelper
import com.anjlab.android.iab.v3.BillingProcessor.IBillingHandler
import com.anjlab.android.iab.v3.BillingProcessor
import com.anjlab.android.iab.v3.TransactionDetails
import com.apps.adrcotfas.goodtime.settings.PreferenceHelper
import android.content.Intent
import com.apps.adrcotfas.goodtime.R
import com.apps.adrcotfas.goodtime.util.Constants
import javax.inject.Inject

class BillingHelper @Inject constructor(context: Context, val preferenceHelper: PreferenceHelper) : IBillingHelper, IBillingHandler {

    private val mBillingProcessor: BillingProcessor = BillingProcessor.newBillingProcessor(
        context,
        context.getString(R.string.licence_key),
        context.getString(R.string.merchant_id), this
    )

    override fun onProductPurchased(productId: String, details: TransactionDetails?) {
        // do nothing here
    }

    override fun onPurchaseHistoryRestored() {
        var found = false
        for (sku in mBillingProcessor.listOwnedProducts()) {
            if (sku == Constants.sku) {
                found = true
                break
            }
        }
        preferenceHelper.setPro(found)
    }

    override fun onBillingError(errorCode: Int, error: Throwable?) {
        // do nothing here
    }

    override fun onBillingInitialized() {
        mBillingProcessor.loadOwnedPurchasesFromGoogle()
        if (mBillingProcessor.isPurchased(Constants.sku)) {
            preferenceHelper.setPro(true)
        }
    }

    override fun refresh() {
        mBillingProcessor.loadOwnedPurchasesFromGoogle()
    }

    override fun release() {
        mBillingProcessor.release()
    }

    override fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        return run {
            mBillingProcessor.loadOwnedPurchasesFromGoogle()
            if (mBillingProcessor.isPurchased(Constants.sku)) {
                preferenceHelper.setPro(true)
            }
            mBillingProcessor.handleActivityResult(requestCode, resultCode, data)
        }
    }

    init {
        mBillingProcessor.initialize()
    }
}